ocaml/xapi/xapi_pool.ml

```diff
@@ -3487,36 +3511,27 @@ let get_updates_handler (req : Http.Request.t) s _ =
             ~self:(Helpers.get_pool ~__context)
             ~doc:"pool.get_updates" ~op:`get_updates (fun () ->
               try
-                let json_str =
-                  Yojson.Basic.to_string
-                    (Repository.get_pool_updates_in_json ~__context ~hosts)
-                in
-                let size = Int64.of_int (String.length json_str) in
-                Http_svr.headers s
-                  (Http.http_200_ok_with_content size ~keep_alive:false ()
-                  @ [Http.Hdr.content_type ^ ": application/json"]
-                  ) ;
-                Unixext.really_write_string s json_str |> ignore
+                http_res (Repository.get_pool_updates_in_json ~__context ~hosts)
               with
               | Api_errors.(Server_error (failure, _)) as e
                 when List.mem failure failures_of_404 ->
                   error "404: can't get updates for pool: %s"
                     (ExnHelper.string_of_exn e) ;
-                  Http_svr.headers s (Http.http_404_missing ())
+                  http_err failure []
               | Api_errors.(Server_error (failure, _)) as e
                 when not (List.mem failure failures_of_404) ->
                   error "getting updates for pool failed: %s"
                     (ExnHelper.string_of_exn e) ;
-                  raise e
+                  http_err failure []
```

Comment:
This is changing the behaviour of the code, isn't it? Would this have wider impact since we no longer raise an exception? I wonder if it would be better to preserve the exception.

Reply:
I think the `raise` in the http handler will return client some 5xx error, which we also need to change to 200 to let client get this detailed error.

Reply:
I think the idea in design is, if it's one of `Api_errors`, it can be encoded into the new error structure. So all of these can be handled in this logic. `failures_of_404` can be removed actually.

Reply:
I don't have the full context but there is a big difference between a HTTP 200 and 500. You can't use 200 to return information about an error - the request failed and that needs to be reflected in the error code.

Reply:
There is a statement in the API reference:

```
When a low-level transport error occurs, or a request is malformed at the HTTP or RPC level,
the server may send an HTTP 500 error response, or the client may simulate the same.
...
All other failures are reported with a more structured error response, to allow better automatic
response to failures, proper internationalisation of any error message, and easier debugging.
```

We don't return 500 if RPCs fail due to xapi throwing an error, we return 200 with the error object in the response. The call `/updates` is implemented as an http backend, but the SDK expects a logic similar to the other RPCs, so currently it does not handle non-200 codes and cannot deduce the specific API error.

Note that this is the case with all the http actions, but I found it particularly problematic in the case of the updates, hence the current ticket. We might want to consider reviewing the way the SDK handles http codes at some point in future, but it would require assessing the impact on the implementing clients first.